upgraded to V4

shunjizhan opened the pull request:
upgraded to version 4.0.0, with some breaking changes. Now it has better code structure and better algorithms.

shunjizhan commented:
I haven't remove some comments yet, will do it later after functionalities work

mboperator commented on src/Components/FolderTree.js:
```diff
@@ -67,37 +71,61 @@ class FolderTree extends Component {
+            path={[]}
+          />
+
+          <FolderToolbar addObj={()=>{}} deleteObj={this.deleteSeletedObj} />
```
Can we add a boolean here for showing/hiding the `FolderToolbar`, similar to `showPane`

shunjizhan committed:
basic webpack setup, can run hello world html

shunjizhan committed:
remove dist folder

shunjizhan committed:
setup basic dev env, hot reloading works

shunjizhan committed:
updated dev server

shunjizhan committed:
restructure scripts and config

shunjizhan committed:
updated HMR that it clears console on each build

shunjizhan committed:
Added Sandbox

Seperated actual folder tree and the sandbox to test it

shunjizhan committed:
Added Icon to Public HTML

- added ico image
- moved public HTML and ico to the new public folder
- edited webpack to add icon

shunjizhan committed:
Set Up FolderTree Skeleton

- added test data
- added skeleton helper functions to initialize tree state

shunjizhan committed:
Added Jest Framework

- installed jest
- add basic jest config so `yarn jest` works
- add skeleton test for utils

shunjizhan committed:
Setup Eslint

- installed and setup basic eslint configs, plugins and rules, so we can now run `yarn lint`.
- fixed a couple lint error

shunjizhan committed:
added airbnb eslint rules and fixed lint errors

shunjizhan committed:
Set Up Husky Pre-commit Pipeline

- installed husky and lint-staged package
- added and configured more eslint rules
- fixed some exlint errors

shunjizhan committed:
run tests in precpush hook

shunjizhan committed:
More Eslint Setup

- setup more rules
- changed yarn lint command so it can lint jsx file along with js files
- lint style fixes

shunjizhan committed:
folderTree can intialzie data with ids

- Implemented addUniqIds function and tests for it
- FolderTree can intialzie data with ids
- more linter rules

shunjizhan committed:
clean up jest config

- clean up unused jest configs
- add test coverage to `yarn test`

shunjizhan committed:
clean up jest config

- clean up unused jest configs
- add test coverage to `yarn test`

shunjizhan committed:
initialized data with checked status

implemented `setCheckedStatus1()` so that data will be initialized with checked status, which defaults to 0

shunjizhan committed:
Merge branch 'v4-upgrade' of github.com:shunjizhan/React-Folder-Tree into v4-upgrade

shunjizhan committed:
able to integrate scss

- added packages to support importing scss
- configured webpack to compile scss

shunjizhan committed:
able to render tree nodes

- added TreeNode component
- folder tree can now render basic tree nodes according to the data
- when (un)check a node, top level onChange will log current tree state
- changed jest threshold to very low (will add more tests later, don't enforce coverage now)

shunjizhan committed:
update eslint to use babel

shunjizhan committed:
added and enabled optional chaining

shunjizhan committed:
Add Checkbox Component

added a checkbox component which should be able to take care of half check (didn't test yet),
this checkbox component is a wrapper of a native checkbox input DOM, it only handles view ,so it only takes a 'status' prop which can be 1, 0, or 0.5 for half check. It also takes a 'onchange' prop which is passed directly to native checkbox DOM.

shunjizhan committed:
half check works!!!

- finished checkNode() function so that it can handle state change for half check.
- refactored some previous functions
- added a couple helper functions getNewCheckStatus(), setStatusDown(), updateStatusUp() for checkNode()
- added unit tests for them

noe we can see the half check works in sandbox!!

shunjizhan committed:
set up jest and enzyme testing env

- added necessary packages
- changed some config files so enzyme and jest tests can run

shunjizhan committed:
added tests for CheckBox component

shunjizhan committed:
refactored code to pass shared util functions using context

- now we don't need to pass same handleCheck() over and over to each level of TreeNode
- it enables to pass future utils more easily, such as handleNameChange()

shunjizhan committed:
adde script to create new component files

shunjizhan committed:
added renameNode() function

- added renameNode() function, which can handle node name change
- adde unit test for it

shunjizhan committed:
added EditableName component

- show an input box, which is an controlled component
- trigger onNameChange function when click ok button
- reset input box when click cancel button
- will replace ok and cancel button with better UI component

shunjizhan committed:
integrated EditableName component to TreeNode

- pass handleNodeRename() to TreeNode in context
- when user type new name in the input box, and  click ok, rename will take effect, and tree state will change accordingly

shunjizhan committed:
added deleteNode function

- now we are able to delete a note by path
- unit tested this function to make sure all the check status are still correct

shunjizhan committed:
integrated deleteNode function to folder tree

- passed deleteNode() as context to TreeNode
- extracted isEditing state from EditableName to upper level TreeNode component to better accompany deleteNode()
- now we are able to select a node, and choose to delete or rename it

shunjizhan committed:
solved a bug for computing state after deleting a node

previously when deleted the last children of a parent node, parent node will be computed to have check status 1.

Solved this bug by checking if children is empty, then return original check status.

shunjizhan committed:
added react-icons package

shunjizhan committed:
Integrated Icons

used react icons replacing previous buttons. Added:
- folder open/close icon
- folder/file icon
- toolbar icons: create, update, delete, cancel, confirm icons

shunjizhan committed:
added functionality to add a node

This function can add a node, which can be either a file or folder.
It will first get current maximum ids, and set a this new node id to be maxId + 1.
Unit tested addNode() and findMaxId()

note that for better UI, files are added to the front of array, folders are to the back

shunjizhan committed:
integrated addNode() to folder tree

Now in folder tree ui we can click button to either add a file or a folder.
This button is only available for folder components

shunjizhan committed:
group all scss to one file

shunjizhan committed:
massive icon UI updates

Wrapped each icon in icon containers, and gave icon className of 'icon' + it's own name.
With these classNames, added hover color to each Icon.

shunjizhan committed:
added tree state viewer in sandbox

shunjizhan committed:
set up sandbox basic structure and envs

- changed components to add ability to provide custom Icons
- added eslintignore to ignore checking for all sandbox files
- separated each different sandbox to a component, and import them to sandbox.jsx, so it can decide which kind of sandbox to render
- added custom Icon sandbox

shunjizhan committed:
added toggle open method in utils

previously we manage the isOpen status in each TreeNode component, but this won't allow user to provide custom open status. So we need to manage isOpen property at the top level tree state.

added toggle open method in utils and unit tested it so it will provide ability to manage isOpen at the top level.

shunjizhan committed:
manage isOpen at the top level tree state

integrated the toggleIsOpen method to the folder tree, so now it manages each isOpen state of node at the whole tree state level.

Note that only folder component, i.e. treenode with 'children' attribute can have isOpen atribute.

shunjizhan committed:
fix the path error for yarn build

shunjizhan committed:
updated main entry in package json to use the production bundle

shunjizhan committed:
production ready

seperated dev and prof webpack config, so in dev mode it will use the sandbox html with the dev-server, while in prod mode it will build a bundle.js in dist/, with only the FolderTree module

shunjizhan committed:
slightly changed webpack setting for production

shunjizhan committed:
changed internal tree state to use _id instead of id

shunjizhan committed:
final polish before publishing

- changed id to _id in the tree state, so people can still have id in the original state
- added api of indentPixels

shunjizhan committed:
updated README for publishing

shunjizhan committed:
fixed tests related to chaning id to _id
